Add a blog post about secure MCP SSE server (#2284)
quarkusio/quarkusio.github.io: _posts/2025-04-28-secure-mcp-sse-server.adoc

---
layout: post
title: 'Getting ready for secure MCP with Quarkus MCP Server'
date: 2025-04-28
tags: ai mcp security
synopsis: 'Explain how MCP clients can access Quarkus MCP SSE servers with access tokens'
author: sberyozkin
---
:imagesdir: /assets/images/posts/secure_mcp_sse_server

== Introduction

https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow.

While it will take a bit of time for the new MCP specification to be widely supported, you can already add authentication to client and server following the https://modelcontextprotocol.io/specification/2024-11-05[previous MCP version].

> Indeed, you can use any MCP client that can receive an access token and pass it to the MCP server.

Reviewer: ```suggestion
You only need an MCP client that can receive an access token and pass it to the MCP server and, obviously, an MCP server that verifies the token.
```

> In this post, we will create a https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP SSE Server] that requires authentication to access its tools.

Author: Just a tiny typo update

=== Implement Login endpoint

Currently, MCP clients can not use the authorization code flow themselves, therefore we implement an OAuth2 login endpoint which will return a GitHub token for the user to use it with MCP clients which can work with bearer tokens.

Add another dependency to support Qute templates:

[source,xml]
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-rest-qute</artifactId> <2>
</dependency>
----

and implement the login endpoint:

[source,java]
----
package org.acme;

import io.quarkus.oidc.AccessTokenCredential;
import io.quarkus.oidc.UserInfo;
import io.quarkus.qute.Template;
import io.quarkus.qute.TemplateInstance;
import io.quarkus.security.Authenticated;
import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;

@Path("/login")
@Authenticated
public class LoginResource {

    @Inject
    UserInfo userInfo; <1>
    
    @Inject
    AccessTokenCredential accessToken; <2>

    @Inject

>     Template accessTokenPage;

Author: @cescoffier Yeah, I link the source at the start of the blog, and it is just a basic Qute template, I thought its content was not really important as we already have quote a few images, but let me link to it in the demo source

    @GET
    @Produces("text/html")
    public TemplateInstance poem() {
        return accessTokenPage.data("name", userInfo.getName()).data("accessToken", accessToken.getToken()); <3>
    }
}
----
<1> GitHub access tokens are binary and Quarkus OIDC indirectly verifies them by using them to request GutHub specific `UserInfo` representation.
<2> `AccessTokenCredential` is used to capture a binary GitHub access token.
<3> After the user logs in to GitHub and is redirected to this endpoint, the access token will be returned to the user in an HTML page generated with Qute.

=== Update the configuration to support GitHub

The <<initial-configuration, configuration>> that was used to run the MCP server in dev mode was suffient because Keycloak Dev Service was supporting the OIDC login.

To work with GitHub in prod mode, we update the configuration as follows:

[source,properties]
----
%prod.quarkus.oidc.provider=github <1>
%prod.quarkus.oidc.application-type=service <2>

> %prod.quarkus.oidc.login.provider=github <3>

Author: @cescoffier Sure, yeah, I've reworked it a bit to try to avoid some upfront references to the config

%prod.quarkus.oidc.login.client-id=github-application-client-id
%prod.quarkus.oidc.login.credentials.secret=github-application-client-secret

quarkus.http.auth.permission.authenticated.paths=/mcp/sse <4>
quarkus.http.auth.permission.authenticated.policy=authenticated
----
<1> Default Quarkus OIDC configuration requires that only GitHub access tokens can be used to access MCP SSE server.
<2> By default, `quarkus.oidc.provider=github` supports an authorization code flow only. `quarkus.oidc.application-type=service` overrides it and requires the use of bearer tokens. 
<3> Use GitHub authorization code flow to support the login endpoint with a dedicated Quarkus OIDC `login` https://quarkus.io/guides/security-openid-connect-multitenancy[tenant] configuration.
<4> Enforce an authenticated access to the main MCP SSE endpoint during the initial handshake. See also how the tool is secured with an annotation in the <<tool>> section above.

[NOTE]
====
Note the use of the `%prod.` prefixes. It ensures the configuration properties prefixed with `%prod.` are only effective in prod mode and not interfering with dev mode. 
====

=== Install and run application

> First, add `quarkus.package.jar.type=uber-jar` to the application.properties.

Author: Hi @cescoffier 

Sure, I believe we use `jbang` to make it easy to plugin MCP servers into Claude and other MCP clients, so, once Claude will start supporting logging in users, this demo will just work there with `jbang`.

But I've added a note:

```
[NOTE]
====
The `quarkus.package.jar.type=uber-jar` property is set to make it possible to run the MCP server with `jbang`.

If you prefer, you can simply package and run the MCP server instead:

[source,shell]
----
mvn package
java -jar target/quarkus-app/quarkus-run.jar
----
====
```

Hope it covers both options

> First, add `quarkus.package.jar.type=uber-jar` to the application.properties.

Reviewer: Claude only supports stdio transport (for now).
So, switch to what Quarkus recommends, not planning on something we don't know.

[[mcp-inspector]]
=== Use MCP Inspector to access MCP server

Launch https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector]:

[source,shell]
----
npx @modelcontextprotocol/inspector
----

Paste the copied GitHub access token to the `Bearer Token` field and connect to the Quarkus MCP SSE server:

image::mcp_inspector_connect.png[MCP Inspector Connect,align="center"]

Next, make a `user-name-provider` tool call:

image::mcp_inspector_tool_call.png[MCP Inspector Tool Call,align="center"]

=== Use curl to access MCP server

Finally, let's use `curl` and also learn a little bit how both the MCP protocol and MCP SSE transport work. 

First, access the main SSE endpoint without the GitHub access token:

[source,shell]
----
curl -v localhost:8080/mcp/sse
----

You will get HTTP 401 error.

Use the access token to access MCP server:

```shell script
curl -v -H "Authorization: Bearer gho_..." localhost:8080/mcp/sse
```

and get an SSE response such as:

[source,shell]
----
< content-type: text/event-stream
< 
event: endpoint
data: /messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
----

The SSE connection is created.

> Now open another window and use the same access token to initialize the curl as MCP client, and access the tool, using the value of the `data` property to build the target URL.

Reviewer: Add something like:

We cannot invoke the tool directly, we first need to follow the MCP handshake protocol:

Initialize the client:

[source,shell]
----
curl -v -H "Authorization: Bearer gho_..." -H "Content-Type: application/json" --data @initialize.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
----

where the `initialize.json` file has a content like this:

[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "initialize",
  "params": {
    "protocolVersion": "2024-11-05",
    "capabilities": {
      "roots": {
        "listChanged": true
      },
      "sampling": {}
    },
    "clientInfo": {
      "name": "CurlClient",
      "version": "1.0.0"
    }
  }
}
----

> and send the initialization notification:

Author: @cescoffier This was done in the previous request, I've tried the text for both requests a little bit

[source,shell]
----
curl -v -H "Authorization: Bearer gho_..." -H "Content-Type: application/json" --data @initialized.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
----

where the `initialized.json` file has a content like this:

```json
{
  "jsonrpc": "2.0",
  "method": "notifications/initialized"
}
```

And call the tool:

[source,shell]
----
curl -v -H "Authorization: Bearer gho_..." -H "Content-Type: application/json" --data @call.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
----

where the `call.json` file has a content like this:

[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 2,
  "method": "tools/call",
  "params": {
    "name": "user-name-provider",
    "arguments": {
    }
  }
}
----

Now look at the SSE connection window and you will see the name from your GitHub account returned.

== Conclusion

> In this blog post, we have explained how you can easily create a secure Quarkus MCP SSE server, obtain an access token and use it to access the MCP server tool in dev mode with `Quarkus MCP SSE Server Dev UI` and prod mode with both the https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the curl tools.

Reviewer: ```suggestion
In this blog post, we explained how you can easily create a authenticated Quarkus MCP SSE server, obtain an access token and use it to access the MCP server tool in dev mode with `Quarkus MCP SSE Server Dev UI` and prod mode with both the https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the curl tools.
You can use any MCP client that allows passing a bearer token to the server.
```

Notice, that there is no real difference in how OAuth2 is done for either Quarkus MCP server or REST endpoints. The most complex part is to get the settings configured correctly for your OAuth2 provider - but when all is done you just apply a few annotations to mark relevant methods as secure and Quarkus handles the authentication for you.

> The Quarkus team is keeping a close eye on the MCP Authorization specification evolution and working on having all possible MCP Authorization scenarios supported.

Reviewer: ```suggestion
This blog post uses the previous version of the MCP protocol. The Quarkus team is keeping a close eye on the new MCP Authorization specification evolution and working on having all possible MCP Authorization scenarios supported.
```

> Stay tuned for more updates !

Reviewer: ```suggestion
Stay tuned for more updates!
```

    
    @Inject
    AccessTokenCredential accessToken; <2>

    @Inject
    Template accessTokenPage;

    @GET
    @Produces("text/html")
    public TemplateInstance poem() {
        return accessTokenPage
           .data("name", userInfo.getName())
           .data("accessToken", accessToken.getToken()); <3>
    }
}
----
<1> GitHub access tokens are binary and Quarkus OIDC indirectly verifies them by using them to request GutHub specific `UserInfo` representation.
<2> `AccessTokenCredential` is used to capture a binary GitHub access token.
<3> After the user logs in to GitHub and is redirected to this endpoint, the access token will be returned to the user in an HTML page generated with a simple https://github.com/quarkiverse/quarkus-mcp-server/tree/main/samples/secure-mcp-sse-server/src/main/resources/templates/accessTokenPage.html[Qute template].
Of course, you would not do that in a real application. It is just an example to demonstrate the capability.

=== Package and run the MCP Server

Package the MCP server application:

[source,shell]
----
mvn package
----

Run it:

[source,shell]
----
java -jar target/quarkus-app/quarkus-run.jar
----

[NOTE]
====

> You can also run the MCP server with `jbang`.

Reviewer 2: ```suggestion
You can also run the MCP server from its Maven coordinates directly with `jbang`.
```

> You have to set the `quarkus.package.jar.type=uber-jar` property first, and then install and run it with `jbang`:

Reviewer 2: ```suggestion
When using Maven coordinates its advantageous to use `quarkus.package.jar.type=uber-jar` property first, and then install and run it with `jbang`:
```

that setting is not needed for jbang.

you can also just run the generated .jar directly with java or jbang.  so adjusted to the correct thing above.

Reviewer 2: actually - you sholdn't even needed to use uber-jar since you are not using @fatjar thus dependencies will be resolved anyway. So its unnecessary - its mvn install that is needed.

Author: @maxandersen Indeed, thanks for the hint, I just dropped a ref to the uber jar property for now.

While it will take a bit of time for the new MCP specification to be widely supported, you can already add authentication to client and server following the https://modelcontextprotocol.io/specification/2024-11-05[previous MCP version].

You only need an MCP client that can receive an access token and pass it to the MCP server and, obviously, an MCP server that verifies the token.

In this post, we will detail how you can enforce authentication with the https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP SSE Server].

We will first use Keycloak as an OpenID Connect (OIDC) provider to login and use a Keycloak JWT access token to access the server with `Quarkus MCP SSE Server Dev UI` in dev mode.

Secondly, we will show how to log in using GitHub OAuth2 and use a GitHub binary access token to access the server in prod mode with both https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the `curl` tools.

[[initial-mcp-server]]
== Step 1 - Create an MCP server using the SSE transport

> First, let's create a secure Quarkus MCP SSE server that requires authentication to establish Server-sent Events (SSE) connection and also when invoking the tools.

Reviewer: ```suggestion
First, let's create a secure Quarkus MCP SSE server that requires authentication to establish Server-Sent Events (SSE) connection and also when invoking the tools.
```

====

After logging in with `Keycloak` as `alice`, copy the acquired access token using a provided copy button:

image::login_and_copy_access_token.png[Login and copy access token,align="center"]

[[mcp-server-devui]]
=== Use Quarkus MCP Server Dev UI to access the MCP server

Make sure to login and copy the access token as explained in the <<oidc-devui>> section above.

Go to http://localhost:8080/q/dev[Dev UI], find the MCP Server card:

image::mcp_server_devui.png[MCP Server in DevUI,align="center"]

Select its `Tools` option and choose to `Call` the `user-name-provider` tool:

image::mcp_server_choose_tool.png[Choose MCP Server tool,align="center"]

Paste the copied Keycloak access token into the Tool's `Bearer token` field, and request a new MCP SSE session:

image::mcp_server_bearer_token.png[MCP Server Bearer token,align="center"]

Make a tool call and get a response which contains the `alice` user name:

image::mcp_server_tool_response.png[MCP Server tool response,align="center"]

All is good in dev mode; it is time to see how it will work in prod mode.

> Before that, make sure you stop the MCP server whuch runs in dev mode.

Reviewer: ```suggestion
Before that, stop the MCP server, which runs in dev mode.
```

----
quarkus.http.auth.permission.authenticated.paths=/mcp/sse <1>
quarkus.http.auth.permission.authenticated.policy=authenticated

%prod.quarkus.oidc.provider=github <2>
%prod.quarkus.oidc.application-type=service <3>

%prod.quarkus.oidc.login.provider=github <4>
%prod.quarkus.oidc.login.client-id=github-application-client-id <5>
%prod.quarkus.oidc.login.credentials.secret=github-application-client-secret <5>
----
<1> Enforce an authenticated access to the main MCP SSE endpoint during the initial handshake. See also how the tool is secured with an annotation in the <<tool>> section above.
<2> Default Quarkus OIDC configuration requires that only GitHub access tokens can be used to access MCP SSE server.
<3> By default, `quarkus.oidc.provider=github` supports an authorization code flow only. `quarkus.oidc.application-type=service` overrides it and requires the use of bearer tokens. 
<4> Use GitHub authorization code flow to support the login endpoint with a dedicated Quarkus OIDC `login` https://quarkus.io/guides/security-openid-connect-multitenancy[tenant] configuration.
<5> Use the client id and secret generated in the <<register-github-application>> section.

[NOTE]
====

> Note the use of the `%prod.` prefixes. It ensures the configuration properties prefixed with `%prod.` are only effective in prod mode and not interfering with dev mode. 

Reviewer: ```suggestion
Note the use of the `%prod.` prefixes. It ensures the configuration properties prefixed with `%prod.` are only effective in prod mode and do not interfere with dev mode. 
```

====

=== Implement Login endpoint

> Currently, MCP clients can not use the authorization code flow themselves, therefore we implement an OAuth2 login endpoint which will return a GitHub token for the user to use it with MCP clients which can work with bearer tokens.

Reviewer: ```suggestion
Currently, MCP clients can not use the authorization code flow themselves. Therefore, we implement an OAuth2 login endpoint that will return a GitHub token for the user to use with MCP clients, which can work with bearer tokens.
```

Launch https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector]:

[source,shell]
----
npx @modelcontextprotocol/inspector
----

Paste the copied GitHub access token to the `Bearer Token` field and connect to the Quarkus MCP SSE server:

image::mcp_inspector_connect.png[MCP Inspector Connect,align="center"]

Next, make a `user-name-provider` tool call:

image::mcp_inspector_tool_call.png[MCP Inspector Tool Call,align="center"]

You will see the name from your GitHub account returned.

=== Use curl to access the MCP server

Finally, let's use `curl` and also learn a little bit how both the MCP protocol and MCP SSE transport work. 

First, access the main SSE endpoint without the GitHub access token:

[source,shell]
----
curl -v localhost:8080/mcp/sse
----

You will get HTTP 401 error.

Use the access token to access MCP server:

```shell script
curl -v -H "Authorization: Bearer gho_..." localhost:8080/mcp/sse
```

and get an SSE response such as:

[source,shell]
----
< content-type: text/event-stream
< 
event: endpoint
data: /messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
----

> The SSE connection is created. 

Reviewer: ```suggestion
The SSE connection is created.  Note the unique path in the received `data`, we need this path to invoke the tools.
```
